cloud/aws/actuators/machine/actuator.go

```diff
@@ -188,6 +248,401 @@ func (a *Actuator) Exists(cluster *clusterv1.Cluster, machine *clusterv1.Machine
+// `deleted` map.
+//
+// Second
+// First, we create a map of tags that have been created or updated by comparing
+// the `src` map to the `annotation`. If entries in the `src` are not in the
+// `annotation`, they are new. If entries in the `src` are in the `annotation`,
+// we compare their values to see if they need updating. If action must be
+// taken, entries are added to the `created` map.
+// We also build up the `newAnnotation` in this phase while building the
+// `create` map.
+//
+// This method is significantly simpler than the securityGroupsChanged method as
+// we only have to compare the `src` to our `annotation`.
+func (a *Actuator) tagsChanged(annotation map[string]interface{}, src map[string]string) (bool, map[string]string, map[string]string, map[string]interface{}) {
+	// Bool tracking if we found any changed state.
+	changed := false
+
+	// Tracking for created/updated
+	created := map[string]string{}
+
+	// Tracking for tags that were deleted.
+	deleted := map[string]string{}
+
+	// The new annotation that we need to set if anything is created/updated.
+	newAnnotation := map[string]interface{}{}
+
+	// Loop over annotation, checking if entries are in src.
+	// If an entry is present in annotation but not src, it has been deleted
+	// since last time. We flag this in the deleted map.
+	for t, v := range annotation {
+		_, ok := src[t]
+
+		// Entry isn't in src, it has been deleted.
+		if !ok {
+			// Cast v to a string here. This should be fine, tags are always
+			// strings.
+			deleted[t] = v.(string)
+			changed = true
+		}
+	}
+
+	// Loop over src, checking for entries in annotation.
+	//
+	// If an entry is in src, but not annotation, it has been created since
+	// last time.
+	//
+	// If an entry is in both src and annotation, we compare their values, if
+	// the value in src differs from that in annotation, the tag has been
+	// updated since last time.
+	//
+	for t, v := range src {
+		av, ok := annotation[t]
+
+		// Entries in the src always need to be noted in the newAnnotation. We
+		// know they're going to be created or updated.
+		newAnnotation[t] = v
+
+		// Entry isn't in annotation, it's new.
+		if !ok {
+			created[t] = v
+			newAnnotation[t] = v
+			changed = true
+			continue
+		}
+
+		// Entry is in annotation, has the value changed?
+		if v != av {
+			created[t] = v
+			changed = true
+		}
+
+		// Entry existed in both src and annotation, and their values were
+		// equal. Nothing to do.
+	}
+
+	// We made it through the loop, and everything that was in src, was also
+	// in dst. Nothing changed.
+	return changed, created, deleted, newAnnotation
+}
+
+// Returns a map[string]interface from a JSON annotation.
+// This method gets the given `annotation` from the `machine` and unmarshalls it
+// from a JSON string into a `map[string]interface{}`.
+func (a *Actuator) machineAnnotationJSON(machine *clusterv1.Machine, annotation string) (map[string]interface{}, error) {
```

Comment:
What is the logic behind using annotations? could querying ec2 replace annotations?

Reply:
Because we need to track the history of tags / sgs that were applied by this controller as opposed to other things. See discussion in [implementers' office hours](https://docs.google.com/document/d/1IZ2-AZhe4r3CYiJuttyciS7bGZTTx4iMppcA8_Pr3xE/edit#heading=h.fb006hyr6m70) and Slack [1](https://kubernetes.slack.com/archives/C8TSNPY4T/p1537883775000100), [2](https://kubernetes.slack.com/archives/C8TSNPY4T/p1537886836000100)

Reply:
As @randomvariable says, we can't easily query AWS for this information as we need to know:
  - The incoming new state that the machine actuator `Update` method is acting on
  - The current state of the Tags and SecurityGroups as AWS knows it (`DescribeInstances` via `InstanceIfExists`)
  - The state as we left it last time the `Update` method was called (the annotation)

From these, we work out what we have been, and what we are currently responsible for, letting us know which Tags or SecurityGroups we can delete if they disappear from the `AdditionalTags` or `AdditionalSecurityGroups` fields. 

This is due to the machine actuator not being responsible for the total state of the machine Tags and SecurityGroups (supporting comments for this are in the [`AWSMachineProviderConfig`](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/8ece4cacc094348080001c4ee6791097c5080308/cloud/aws/providerconfig/types.go#L27) type.